### ISSUE TITLE
allow() fails when other warnings trigger (possibly on macro)

### ISSUE THREAD

**Issue by xmo-odoo** [rust-lang/rust-clippy#10355]
### Summary

When a statement (macro?) triggers two different warnings and one of them is `allow`-ed, the allow-ed warning triggers anyway

### Reproducer

I tried this code:

```rust
    #[allow(clippy::write_with_newline)]
    write!(w, "{}\n", val);
```

I saw this happen:

```
warning: unused attribute `allow`
  --> src/lib.rs:9:5
   |
9  |     #[allow(clippy::write_with_newline)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the built-in attribute `allow` will be ignored, since it's applied to the macro invocation `write`
  --> src/lib.rs:10:5
   |
10 |     write!(w, "{}\n", val);
   |     ^^^^^
   = note: `#[warn(unused_attributes)]` on by default

warning: using `write!()` with a format string that ends in a single newline
  --> src/lib.rs:10:5
   |
10 |     write!(w, "{}\n", val);
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#write_with_newline
   = note: `#[warn(clippy::write_with_newline)]` on by default
help: use `writeln!` instead
   |
10 -     write!(w, "{}\n", val);
10 +     writeln!(w, "{}", val);
   |

warning: unused `std::result::Result` that must be used
  --> src/lib.rs:10:5
   |
10 |     write!(w, "{}\n", val);
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this `Result` may be an `Err` variant, which should be handled
   = note: `#[warn(unused_must_use)]` on by default
   = note: this warning originates in the macro `write` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I expected to see this happen:
```
warning: unused `std::result::Result` that must be used
  --> src/lib.rs:10:5
   |
10 |     write!(w, "{}\n", val);
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this `Result` may be an `Err` variant, which should be handled
   = note: `#[warn(unused_must_use)]` on by default
   = note: this warning originates in the macro `write` (in Nightly builds, run with -Z macro-backtrace for more info)
```

The version which does not trigger `unused_must_use`:
```rust
    #[allow(clippy::write_with_newline)]
    write!(w, "{}\n", val).unwrap();
```
does not trigger any warning

### Version

```text
rustc 1.67.0 (fc594f156 2023-01-24)
binary: rustc
commit-hash: fc594f15669680fa70d255faec3ca3fb507c3405
commit-date: 2023-01-24
host: x86_64-unknown-linux-gnu
release: 1.67.0
LLVM version: 15.0.6

Also current playground
```

**Comment by Alexendoo** [comment]
It's because attributes applied to macros invocations are ignored, here's a non clippy example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=73c25a903e3242ea1ed8e41f7d6277a5

The `write!(w, "{}\n", val).unwrap()` example contains a macro invocation but isn't a macro invocation statement itself, so the attribute does apply there

I would guess the reason is related to attributes not being allowed on expressions (https://github.com/rust-lang/rust/issues/15701), or that a macro invocation in statement position can expand to multiple statements
